### PR TITLE
[SEMVER-MAJOR] throw error for undefined mixin

### PR DESCRIPTION
**Pull request by alexpit**
connect to #566

### CODE REVIEW

**Comment by alexpit** [comment]
@richardpringle

**Comment by alexpit** [comment]
@slnode test please

**Comment by alexpit** [comment]
@bajtos

**Comment by alexpit** [comment]
connect to #566

**Comment by richardpringle** [comment]
@alexpit, don't forget to update the release notes!

File: lib/mixins.js [hunk 1]
```diff
@@ -45,6 +45,7 @@ MixinProvider.prototype.applyMixin = function applyMixin(modelClass, name, optio
       modelClass.mixin(model, options);
     } else {
       debug('Invalid mixin: %s', name);
+      throw new Error('Invalid mixin: ' + name);
```
Comment (bajtos) [inline lib/mixins.js]: Please include the model name in the error message, otherwise it's impossible to find the model which is referencing this missing mixing.

I would also reword the message a bit, e.g.

```js
'Model ' + modelClass.modelName + ' uses unknown mixin ' + name
```

File: test/mixins.test.js [hunk 2]
```diff
@@ -79,11 +79,24 @@ describe('Model class', function() {
     properties.city.should.eql({ type: String, required: true });
   });
 
+  it('should fail to apply an undefined mixin class', function(done) {
+    var memory = new DataSource('mem', { connector: Memory }, modelBuilder);
+    function applyMixin() {
+      var UndefinedMixin;
```
Comment (bajtos) [inline test/mixins.test.js]: Unused variable, please remove.

File: test/mixins.test.js [hunk 3]
```diff
@@ -110,6 +123,25 @@ describe('Model class', function() {
     });
   });
 
+  it('should fail to apply undefined mixin', function(done) {
+    var memory = new DataSource('mem', { connector: Memory }, modelBuilder);
+    var Item = memory.createModel('Item', { name: 'string' }, {
+      mixins: {
```
Comment (bajtos) [inline test/mixins.test.js]: These mixins are not needed by this test, please remove the entire `mixins:{}` section.

**Comment by bajtos** [comment]
@alexpit thank you for the pull request, you are headed in the right direction! I left few comments to address, see above.

File: test/mixins.test.js (hunk 2 above)
Comment (alexpit) [inline test/mixins.test.js]: If you're referring to the UndefinedMixin, that's used, otherwise please clarify. Thanks

**Comment by alexpit** [comment]
@richardpringle, please review. thx

**Comment by alexpit** [comment]
@bajtos, please review addressed comments. thx

File: 3.0-RELEASE-NOTES.md [hunk 4]
```diff
@@ -49,3 +49,12 @@ without any callback, a promise is returned.
...
 for more details.
+
+## Applying an undefined mixin to a LoopBack model throws an error
+
+When applying an undefined mixin to a LoopBack model, a warning message was
+logged regarding the invalid mixin, which needs to be addressed rather than
+silently ignored. This has been fixed in 3.0, therefore an undefined mixin
+applied to a LoopBack model will throw an error that needs to be handled.
+
+Please see [related code change](https://github.com/strongloop/loopback-datasource-juggler/pull/944)
```
Comment (richardpringle) [inline 3.0-RELEASE-NOTES.md]: Nit pick:

`Please see [related code change]() for details.`

File: lib/mixins.js [hunk 5]
```diff
@@ -41,10 +41,12 @@ MixinProvider.prototype.applyMixin = function applyMixin(modelClass, name, optio
     // Try model name
     var model = this.modelBuilder.getModel(name);
     if (model) {
-      debug('Mixin is resolved to a model: %s', name);
+      debug('Mixin "' + name + '" resolves to model: %s', modelClass.modelName);
```
Comment (bajtos) [inline lib/mixins.js]: This is not correct. Mixin "name" is resolved to a model called "name" in this branch, and then it is applied to another model called "modelClass.modelName".

It's better to avoid making unrelated code changes, as it makes code archaeology more difficult.

Please revert this line.

File: test/mixins.test.js [hunk 6]
```diff
@@ -79,11 +79,23 @@ describe('Model class', function() {
     properties.city.should.eql({ type: String, required: true });
   });
 
+  it('should fail to apply an undefined mixin class', function(done) {
```
Comment (bajtos) [inline test/mixins.test.js]: Since this test completes on the same turn of the event loop, there is no need to use `done` callback.
Instead of 

```js
it('...', function(done) {
  var memory = ...;
  // ...
  should...
  done();
});
```

write simply:

```js
it('...', function() {
  var memory = ...;
  // ...
  should...
});
```

File: test/mixins.test.js [hunk 7]
```diff
@@ -110,6 +122,17 @@ describe('Model class', function() {
     });
   });
 
+  it('should fail to apply undefined mixin', function(done) {
```
Comment (bajtos) [inline test/mixins.test.js]: ditto, `done` is not needed, please remove

**Comment by alexpit** [comment]
@bajtos, please review addressed comments. thx

**Comment by bajtos** [comment]
LGTM 👏 

Please squash the commits before landing, the commands are `git rebase -i master` and `git push -f`.

**Comment by Amir-61** [comment]
Hey @bajtos and @alexpit,

@bajtos Does Alex need to backprot this change?

@alexpit For `loopback-datasource-juggler`, we may need to backport the changes to `2.x` branch as well; please see [our team documentation](https://github.com/strongloop-internal/loopback-knowledge-base/blob/master/backporting-policy.md). If @bajtos confirmed that you need to backport the changes, please let me know if you need any help for backproting and I can walk you through the steps.

Thanks :-)

**Comment by richardpringle** [comment]
Woah, dependent loopback failed, that can't be good. Also, @alexpit, i'll help you with squashing. 

@bajtos, git push -f? That will force push master. Should be git push origin +branch_name no?

**Comment by Amir-61** [comment]
>@bajtos, git push -f? That will force push master. Should be git push origin +branch_name no?

I believe Miroslav meant run `git push -f ` when you are in your feature branch and after rebasing and squashing the commits in feature branch.

**Comment by richardpringle** [comment]
I believe that depends on your push policy configuration. It's safer to explicitly push a certain branch with '+' causing the force push.

**Comment by bajtos** [comment]
> Does Alex need to backprot this change?

No, we cannot back-port breaking changes.

>  > > git push -f? That will force push master. Should be git push origin +branch_name no?

> > I believe Miroslav meant run git push -f when you are in your feature branch and after rebasing and squashing the commits in feature branch.

> I believe that depends on your push policy configuration. It's safer to explicitly push a certain branch with '+' causing the force push.

Interesting. I think it's rather unlikely that one would one end up with a feature branch configured to be pushed to a long-lived remote branch like `master`, but then there is no harm in being cautious.

BTW in some repositories, we have locked long-lived branches like `master` and `2.x` to reject force-pushes. If you find a repository that does not have `master` locked, then feel free to make the change yourself.

**Comment by bajtos** [comment]
@slnode test please

**Comment by alexpit** [comment]
@bajtos please review

**Comment by bajtos** [comment]
>  PR Linter — PR out of date, needs to be rebased

@alexpit could you please rebase on top of the current master?

**Comment by alexpit** [comment]
@slnode test please
